miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/rest/BoxRestController.java

```diff
@@ -252,8 +290,293 @@ private static BoxableId parseEntityIdentifier(String identifier) {
+   */
+  @RequestMapping(value = "/rest/box/{boxId}/scan/", method = RequestMethod.POST)
+  public @ResponseBody ScanResultsDto getBoxScan(@PathVariable(required = true) int boxId,
+      @RequestBody(required = true) ScanRequest requestData) {
+    try {
+      BoxScanner boxScanner = boxScanners.get(requestData.getScannerName());
+      if (boxScanner == null) {
+        throw new RestException("Invalid scanner specified", Status.BAD_REQUEST);
+      }
+      BoxScan scan = boxScanner.getScan();
+      if (scan == null) {
+        throw new RestException("The scanner did not detect a box!", Status.CONFLICT);
+      }
+
+      Map<String, String> barcodesByPosition = scan.getBarcodesMap();
+      // Extract the valid barcodes and build a barcode to item map
+      Set<String> validBarcodes = barcodesByPosition.values().stream()
+          .filter(position -> hasRealBarcode(scan, position)).collect(Collectors.toSet());
+      Map<String, BoxableView> boxablesByBarcode = boxService.getViewsFromBarcodeList(validBarcodes).stream()
+          .collect(Collectors.toMap(BoxableView::getIdentificationBarcode, Function.identity()));
+
+      // For all the valid barcodes, build a list of DTOs with the updated positions
+      List<BoxableDto> items = barcodesByPosition.entrySet().stream()
+          .filter(entry -> hasRealBarcode(scan, entry.getKey()) && boxablesByBarcode.containsKey(entry.getValue()))
+          .map(entry -> {
+            BoxableDto dto = Dtos.asDto(boxablesByBarcode.get(entry.getValue()));
+            dto.setCoordinates(entry.getKey());
+            return dto;
+          }).collect(Collectors.toList());
+
+      // Collect all the errors
+      List<ErrorMessage> errors = new ArrayList<>();
+
+      // If there's a barcode that wasn't found in the DB, create an error.
+      barcodesByPosition.entrySet().stream()
+          .filter(entry -> hasRealBarcode(scan, entry.getKey()) && !boxablesByBarcode.containsKey(entry.getValue())).map(entry -> {
+            ErrorMessage dto = new ErrorMessage();
+            dto.setCoordinates(entry.getKey());
+            dto.setMessage("Barcode " + entry.getValue() + " not found.");
+            return dto;
+          }).forEachOrdered(errors::add);
+      // If there was a read error, produce an error.
+      scan.getReadErrorPositions().stream().map(position -> {
+        ErrorMessage dto = new ErrorMessage();
+        dto.setCoordinates(position);
+        dto.setMessage("Cannot read tube.");
+        return dto;
+      }).forEachOrdered(errors::add);
+
+      long totalBarcodes = barcodesByPosition.values().stream().filter(position -> hasRealBarcode(scan, position)).count();
+      if (validBarcodes.size() != totalBarcodes) {
+        ErrorMessage dto = new ErrorMessage();
+        dto.message = "Duplicate barcodes detected!";
+        errors.add(dto);
+      }
+
+      // Build the diffs for this box
+      List<DiffMessage> diffs = new ArrayList<>();
+      Box box;
+      try {
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Extract this nested try block into a separate method. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1141)